pkg/object/tcpproxy/connection.go

```diff
@@ -0,0 +1,441 @@
+				return
+			}
+			c.appendBuffer(buf)
+		OUTER:
+			for i := 0; i < 8; i++ {
+				select {
+				case buf, ok := <-c.writeBufferChan:
+					if !ok {
+						return
+					}
+					c.appendBuffer(buf)
+				default:
+					break OUTER
+				}
+			}
+
+			_ = c.rawConn.SetWriteDeadline(time.Now().Add(15 * time.Second))
+			_, err = c.doWrite()
+		}
+
+		if err != nil {
+			if err == iobufferpool.ErrEOF {
+				logger.Debugf("tcp connection local close with eof, local addr: %s, remote addr: %s",
+					c.localAddr.String(), c.remoteAddr.String())
+				_ = c.Close(NoFlush, LocalClose)
+			} else {
+				logger.Errorf("tcp connection error on write, local addr: %s, remote addr: %s, err: %+v",
+					c.localAddr.String(), c.remoteAddr.String(), err)
+			}
+
+			if te, ok := err.(net.Error); ok && te.Timeout() {
+				_ = c.Close(NoFlush, OnWriteTimeout)
+			}
+			//other write errs not close connection, because readbuffer may have unread data, wait for readloop close connection,
+			return
+		}
+	}
+}
+
+func (c *Connection) appendBuffer(buf *iobufferpool.StreamBuffer) {
+	if buf == nil {
+		return
+	}
+	c.ioBuffers = append(c.ioBuffers, buf)
+	c.writeBuffers = append(c.writeBuffers, buf.Bytes())
+}
+
+// Close connection close function
+func (c *Connection) Close(ccType CloseType, event ConnectionEvent) (err error) {
+	defer func() {
+		if r := recover(); r != nil {
+			logger.Errorf("tcp connection close panic, err: %+v\n%s", r, string(debug.Stack()))
+		}
+	}()
+
+	if ccType == FlushWrite {
+		_ = c.Write(iobufferpool.NewEOFStreamBuffer())
+		return nil
+	}
+
+	if !atomic.CompareAndSwapUint32(&c.closed, 0, 1) {
+		return nil
+	}
+
+	// connection failed in client mode
+	if c.rawConn == nil || reflect.ValueOf(c.rawConn).IsNil() {
+		return nil
+	}
+
+	// close tcp rawConn read first
+	if tconn, ok := c.rawConn.(*net.TCPConn); ok {
+		logger.Debugf("tcp connection closed, local addr: %s, remote addr: %s, event: %s",
+			c.localAddr.String(), c.remoteAddr.String(), event)
+		_ = tconn.CloseRead()
+	}
+
+	// close rawConn recv, then notify read/write loop to exit
+	close(c.connStopChan)
+	_ = c.rawConn.Close()
+	c.lastBytesSizeRead, c.lastWriteSizeWrite = 0, 0
+
+	logger.Debugf("tcp connection closed, local addr: %s, remote addr: %s, event: %s", c.localAddr.String(), c.remoteAddr.String(), event)
+
+	if c.onClose != nil {
+		c.onClose(event)
+	}
+	return nil
+}
+
+func (c *Connection) doReadIO() (bufLen int, err error) {
+	if c.readBuffer == nil {
+		c.readBuffer = iobufferpool.TCPBufferPool.Get().([]byte)
+	}
+
+	// add read deadline setting optimization?
+	// https://github.com/golang/go/issues/15133
+	_ = c.rawConn.SetReadDeadline(time.Now().Add(15 * time.Second))
+	return c.rawConn.(io.Reader).Read(c.readBuffer)
+}
+
+func (c *Connection) doWrite() (int64, error) {
+	bytesSent, err := c.doWriteIO()
+	if err != nil && atomic.LoadUint32(&c.closed) == 1 {
+		return 0, nil
+	}
+
+	if bytesBufSize := c.writeBufLen(); bytesBufSize != c.lastWriteSizeWrite {
```

Comment:
what if `c.closed` becomes 1 here?
please be very very careful with `atomic` functions, it is much easier to make mistakes than looks.